interfaces/builtin/posix_mq.go

```diff
@@ -0,0 +1,331 @@
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) validatePermissionList(perms []string, name string) error {
+	for _, perm := range perms {
+		if !strutil.ListContains(posixMQPlugPermissions, perm) {
+			return fmt.Errorf("posix-mq slot permission \"%s\" not valid, must be one of %v", perm, posixMQPlugPermissions)
+		}
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) validatePermissionsAttr(permsAttr interface{}) ([]string, error) {
+	var perms []string
+
+	// Ensure that the permissions attribute is a list
+	if permsList, ok := permsAttr.([]interface{}); ok {
+		// Ensure that each permission in the list is a string
+		for _, i := range permsList {
+			if perm, ok := i.(string); ok {
+				perms = append(perms, perm)
+			} else {
+				return nil, fmt.Errorf(`each posix-mq slot permission must be a string, not %v`, permsAttr)
+			}
+		}
+	} else {
+		return nil, fmt.Errorf(`posix-mq slot "permissions" attribute must be a list of strings, not %v`, permsAttr)
+	}
+
+	return perms, nil
+}
+
+func (iface *posixMQInterface) getPermissions(attrs interfaces.Attrer, name string) ([]string, error) {
+	var perms []string
+
+	if permsAttr, isSet := attrs.Lookup("permissions"); isSet {
+		if permsList, err := iface.validatePermissionsAttr(permsAttr); err != nil {
+			return nil, err
+		} else {
+			perms = permsList
+		}
+	} else {
+		// If the permissions have not been specified, use the defaults
+		perms = posixMQDefaultPlugPermissions
+	}
+
+	if err := iface.validatePermissionList(perms, name); err != nil {
+		return nil, err
+	}
+
+	return perms, nil
+}
+
+func (iface *posixMQInterface) getPath(attrs interfaces.Attrer, name string) (string, error) {
+	var path string
+
+	if pathAttr, isSet := attrs.Lookup("path"); isSet {
+		if pathStr, ok := pathAttr.(string); ok {
+			path = pathStr
+		} else {
+			return "", fmt.Errorf(`posix-mq slot "path" attribute must be a string, not %v`, pathAttr)
+		}
+	} else {
+		return "", fmt.Errorf(`posix-mq slot requires the "path" attribute`)
+	}
+
+	// Path must begin with a /
+	if path[0] != '/' {
+		path = "/" + path
+	}
+
+	if err := iface.validatePath(name, path); err != nil {
+		return "", err
+	}
+
+	return path, nil
+
+}
+
+func (iface *posixMQInterface) validatePath(name, path string) error {
+	if !posixMQNamePattern.MatchString(path) {
+		return fmt.Errorf(`posix-mq "path" attribute must conform to the POSIX message queue name specifications (see "man mq_overview"): %v`, path)
+	}
+
+	if err := apparmor_sandbox.ValidateNoAppArmorRegexp(path); err != nil {
+		return fmt.Errorf(`posix-mq "path" attribute is invalid: %v"`, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`posix-mq "path" attribute is not a clean path: %v"`, path)
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	if err := iface.checkPosixMQAppArmorSupport(); err != nil {
+		return err
+	}
+
```

Comment:
this needs to validate posix-mq and default it to the plug name if it's missing (see similar code for content.go BeforePreparePlug)